src/toil/provisioners/aws/awsProvisioner.py

```diff
@@ -0,0 +1,511 @@
+
+    @classmethod
+    def _createSecurityGroup(cls, ctx, name):
+        # security group create/get. ssh + all ports open within the group
+        try:
+            web = ctx.ec2.create_security_group(name, 'Toil appliance security group')
+        except EC2ResponseError as e:
+            if e.status == 400 and 'already exists' in e.body:
+                pass  # group exists- nothing to do
+            else:
+                raise
+        else:
+            for attempt in retryBoto(retryWhile=groupNotFound, retryFor=300):
+                with attempt:
+                    # open port 22 for ssh-ing
+                    web.authorize(ip_protocol='tcp', from_port=22, to_port=22, cidr_ip='0.0.0.0/0')
+            for attempt in retryBoto(retryWhile=groupNotFound, retryFor=300):
+                with attempt:
+                    # the following authorizes all port access within the web security group
+                    web.authorize(ip_protocol='tcp', from_port=0, to_port=65535, src_group=web)
+            for attempt in retryBoto(retryWhile=groupNotFound, retryFor=300):
+                with attempt:
+                    # open port 5050-5051 for mesos web interface
+                    web.authorize(ip_protocol='tcp', from_port=5050, to_port=5051, cidr_ip='0.0.0.0/0')
+
+    @classmethod
+    def _getProfileARN(cls, ctx):
+        roleName = 'toil'
+        policy = dict(iam_full=iam_full_policy, ec2_full=ec2_full_policy,
+                      s3_full=s3_full_policy, sbd_full=sdb_full_policy)
+        iamRoleName = ctx.setup_iam_ec2_role(role_name=roleName, policies=policy)
+
+        try:
+            profile = ctx.iam.get_instance_profile(iamRoleName)
+        except BotoServerError as e:
+            if e.status == 404:
+                profile = ctx.iam.create_instance_profile(iamRoleName)
+                profile = profile.create_instance_profile_response.create_instance_profile_result
+            else:
+                raise
+        else:
+            profile = profile.get_instance_profile_response.get_instance_profile_result
+        profile = profile.instance_profile
+        profile_arn = profile.arn
+
+        if len(profile.roles) > 1:
+                raise RuntimeError('Did not expect profile to contain more than one role')
+        elif len(profile.roles) == 1:
+            # this should be profile.roles[0].role_name
+            if profile.roles.member.role_name == iamRoleName:
+                return profile_arn
+            else:
+                ctx.iam.remove_role_from_instance_profile(iamRoleName,
+                                                          profile.roles.member.role_name)
+        for attempt in retryBoto(retryWhile=addRoleErrors):
+            with attempt:
+                ctx.iam.add_role_to_instance_profile(iamRoleName, iamRoleName)
+        return profile_arn
+
+
+def retryBoto(retryWhile, retryFor=180, retryAfter=5):
```

Comment:
@JakeNarkizian and I recently generalized the retry logic and moved it to BPL. CGCloud and Toil now use it. You should be able to reimplement `retryBoto()` in terms of BPL's `retry()`. Note that that will also introduce incremental back-off by default, unless you customize the `delays` argument.